File: 2. Add Two Numbers/addTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    private ListNode addTwoNumbersHelper(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null) {
            if (carry != 0) {
                return new ListNode(carry);
            } else {
                return null;
            }
        } 
        
        int l2Val = l2 == null ? 0 : l2.val;
        int l1Val = l1 == null ? 0 : l1.val;
        int sum = l1Val + l2Val + carry;
        
        int newCarry = sum / 10;
        int remains = sum % 10;
        final ListNode node = new ListNode(remains);
        node.next = addTwoNumbersHelper(l1 == null ? null : l1.next, l2 == null ? null : l2.next, newCarry);
        return node;
    }
    
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return addTwoNumbersHelper(l1, l2, 0);
    }
}

File: 2. Add Two Numbers/addTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbersHelper(self, l1, l2, carry):
        if l1 is None and l2 is None and carry == 0:
            return None
        l1Val = l1.val if l1 != None else 0
        l2Val = l2.val if l2 != None else 0
        val = l1Val + l2Val + carry
        intPart = val / 10
        remainder = val % 10
        node = ListNode(remainder)
        node.next = self.addTwoNumbersHelper(l1.next if l1 != None else None, l2.next if l2 != None else None, intPart)
        return node
    
    def addTwoNumbers(self, l1, l2):
        """
            :type l1: ListNode
            :type l2: ListNode
            :rtype: ListNode
            """
        return self.addTwoNumbersHelper(l1, l2, 0)
